zeebe/atomix/cluster/src/main/java/io/atomix/raft/impl/RaftContext.java

```diff
@@ -923,11 +935,30 @@ public long getFirstCommitIndex() {
    *
    * @param firstCommitIndex The first commit index.
    */
-  public void setFirstCommitIndex(final long firstCommitIndex) {
+  public void setFirstCommitIndex(final long firstCommitIndex, final long lastFlushedIndex) {
     if (this.firstCommitIndex == 0) {
       if (firstCommitIndex == 0) {
         return;
       }
+      // The previous leader may have crashed before updating the commit index on current leader.
+      // However, the data was already committed if the current leader had already flushed the
+      // message
+      // in its local storage.
+      // However, if the current leader has not received that message, we may be at risk of losing
+      // data.
+      if (firstCommitIndex < commitIndex && lastFlushedIndex < commitIndex) {
+        final var errorMessage =
+            String.format(
+                """
+                   A majority of nodes has lost committed data: firstCommitIndex(%d), lastflushedIndex(%d) < commitIndex(%d). \
+                   This node will become inactive to avoid overwriting previously committed data, \
+                   but the leader have formed a quorum and will continue to commit new events, \
+                   creating an inconsistent timeline of events: \
+                   THE CLUSTER SHOULD BE STOPPED IMMEDIATELY to further prevent inconsistencies.""",
```

Comment:
🔧 Following our error guidelines https://github.com/camunda/camunda/wiki/Error-Guidelines

```suggestion
                """
                   Expected to set first commit position to after restart, but firstCommitIndex(%d), lastflushedIndex(%d) < commitIndex(%d). \
                   While commitIndex is the last committed index, persisted in the metadata file. \
                   This means a majority of nodes has lost committed data: \
                   This node will become inactive to avoid overwriting previously committed data, \
                   but the leader have formed a quorum and will continue to commit new events, \
                   creating an inconsistent timeline of events: \
                   THE CLUSTER SHOULD BE STOPPED IMMEDIATELY to further prevent inconsistencies.""",
```